src/search/parser/syntax_analyzer.cc

```diff
@@ -0,0 +1,257 @@
+        message << endl
+                << all_tokens << endl
+                << string(all_tokens.size() - remaining_tokens.size(), ' ')
+                << "^";
+        return message.str();
+    }
+};
+
+static ASTNodePtr parse_node(TokenStream &tokens, SyntaxAnalyzerTraceback &traceback);
+
+static void parse_argument(TokenStream &tokens,
+                           vector<ASTNodePtr> &positional_arguments,
+                           unordered_map<string, ASTNodePtr> &keyword_arguments,
+                           SyntaxAnalyzerTraceback &traceback) {
+    if (tokens.has_tokens(2)
+        && tokens.peek(0).type == TokenType::IDENTIFIER
+        && tokens.peek(1).type == TokenType::EQUALS) {
+        string argument_name = tokens.pop().content;
+        tokens.pop(TokenType::EQUALS);
+        if (keyword_arguments.count(argument_name)) {
+            throw ParserError(
+                      "Multiple definitions of the same keyword argument '" +
+                      argument_name + "'.",
+                      traceback);
+        }
+        keyword_arguments[argument_name] = parse_node(tokens, traceback);
+    } else {
+        if (!keyword_arguments.empty()) {
+            throw ParserError(
+                      "Positional arguments have to be defined before any "
+                      "keyword arguments.",
+                      traceback);
+        }
+        positional_arguments.push_back(parse_node(tokens, traceback));
+    }
+}
+
+static ASTNodePtr parse_let(TokenStream &tokens, SyntaxAnalyzerTraceback &traceback) {
+    traceback.push("Let");
+    tokens.pop(TokenType::LET);
+    tokens.pop(TokenType::OPENING_PARENTHESIS);
+
+    traceback.push("Parsing variable name");
+    string variable_name = tokens.pop(TokenType::IDENTIFIER).content;
+    traceback.pop();
+    traceback.push("Found variable name '" + variable_name + "'");
+
+    traceback.push("Expecting separation of variable name and definition");
+    tokens.pop(TokenType::COMMA);
+    traceback.pop();
+
+    traceback.push("Parsing variable definition");
+    ASTNodePtr variable_definition = parse_node(tokens, traceback);
+    traceback.pop();
+
+    traceback.push("Expecting separation of variable definition and expression");
+    tokens.pop(TokenType::COMMA);
+    traceback.pop();
+
+    traceback.push("Parsing expression");
+    ASTNodePtr nested_value = parse_node(tokens, traceback);
+    traceback.pop();
+
+    traceback.push("Expecting closing of let");
+    tokens.pop(TokenType::CLOSING_PARENTHESIS);
+    traceback.pop();
+    traceback.pop();
+    traceback.pop();
+    return utils::make_unique_ptr<LetNode>(
+        variable_name, move(variable_definition), move(nested_value));
+}
+
+static void parse_sequence(
+    TokenStream &tokens, SyntaxAnalyzerTraceback &traceback,
+    TokenType terminal_token, function<void(void)> func) {
+    traceback.push("Parsing arguments");
```

Comment:
This is indeed hard to follow. The problem with `push()`/`pop()` is that if is sometimes tricky to find a good place for the `pop()` (for example, when returning from a function). An alternative that we discussed is similar to
`utils::language::TraceBlock`, an object that is created to start the block and when it goes out of scope, it closes the block in its destructor. Maybe we should change everything to this way of tracing.

Reply:
Instead of
```C++
context.push("whatever");
do_something();
context.pop();
```
it would be
```C++
{
    TraceBlock b = context.trace_block("whatever");
    do_something();
}
```

And instead of 

```C++
context.push("whatever");
if (foo) {
    context.pop();
    return nullptr;
}
const T &value = create_value();
context.pop();
return value;
```
it would be
```C++
{
    TraceBlock b = context.trace_block("whatever");
    if (foo) {
        return nullptr;
    }
    return create_value();
}
```